src/configuration.rs

```diff
@@ -0,0 +1,138 @@
+use serde::{Serialize, Deserialize};
+use std::fs;
+use std::error::Error;
+use std::env;
+
+/// Represents the main configuration for a machine.
+/// 
+/// Fields:
+/// - `mmu_mode`: Optional string that specifies the MMU mode, which will be manually converted to a specific type.
+/// - `memspaces`: A vector of `MemorySpace` that defines different memory regions.
+/// - `dumpfile`: Optional string specifying the path to a dump file.
+/// - `outfile`: Optional string specifying the path to an output file.
+/// 
+/// Description:
+/// - This struct is used to parse and hold the configuration data of a machine from various file formats.
+/// - It includes memory spaces, MMU mode, and optional paths for dump and output files.
+/// 
+/// Purpose:
+/// - To provide a structured representation of machine configuration that can be easily loaded, manipulated, and accessed.
+///
+/// Technical Explanation:
+/// - The `mmu_mode` is stored as a string to allow flexibility in how it is represented and used in different contexts.
+/// - `memspaces` is a vector of `MemorySpace`, which details the type and address range of memory spaces.
+#[derive(Debug, Serialize, Deserialize)]
+struct MachineConfig {
+    mmu_mode: Option<String>,
+    memspaces: Vec<MemorySpace>,
+    dumpfile: Option<String>,
+    outfile: Option<String>,
+}
+
+/// Defines a memory space with a type and address range.
+#[derive(Serialize, Deserialize, Debug)]
+struct MemorySpace {
+    space_type: SpaceType,
+    start_address: u64,
+    end_address: u64,
+}
+
+/// Enumerates types of memory spaces.
+#[derive(Serialize, Deserialize, Debug)]
+enum SpaceType {
+    RAM,
+    ROM,
+}
+
+/// Enumerates RISC-V MMU modes.
+#[derive(Debug, Serialize, Deserialize)]
+enum RiscVMMUMode {
+    SV39,
+    SV48,
+}
+
+/// Loads configuration from a specified file.
+/// 
+/// Args:
+/// - `file_path`: The path to the configuration file.
+/// 
+/// Returns:
+/// - A `Result` containing `MachineConfig` or an error.
+/// 
+/// Description:
+/// - This function reads a configuration file and deserializes it into a `MachineConfig` instance.
+/// 
+/// Purpose:
+/// - To abstract the details of loading and parsing configuration files into a single reusable function.
+///
+/// Technical Explanation:
+/// - The function determines the file format based on the extension and uses the appropriate Serde deserializer.
+/// - It handles TOML format.
+///
+/// Example:
+/// ```rust
+/// let config = load_config("config.toml").unwrap();
+/// println!("{:?}", config);
+/// ```
+fn load_config(file_path: &str) -> Result<MachineConfig, anyhow::Error> {
+    let config_str = fs::read_to_string(file_path)
+        .with_context(|| format!("Failed to read configuration file: {}", file_path))?;
+    let config: MachineConfig = match file_path.rsplit('.').next().context("No file extension found")? {
+        "toml" => toml::from_str(&config_str)
```

Comment:
No validation whatsoever for the loaded content ? What if a specified MMU mode doesn't exists ?